### PR TITLE
GBDK: improve detection accuracy of versions 1.0 - 2020.3.2.0

### PR DESCRIPTION
**Pull request by asiekierka**
This has bothered me for ages, so here's my improvement proposal - complete with test cases.

There may be further ways to distinguish these versions, but I think this granularity is quite satisfactory.

### CODE REVIEW

File: src/entry_names_gbdk.h [hunk 1]
```diff
@@ -14,7 +14,12 @@
     // Tool
     DEF_NAME_STR(STR_GBDK, "GBDK");
     // Versions
-    DEF_NAME_STR(STR_GBDK_2_x_to_2020_3_2_0,       "2.x - 2020.3.2.0");
+    DEF_NAME_STR(STR_GBDK_1_x,                     "1.x");
+    DEF_NAME_STR(STR_GBDK_2_0_x_to_2_0_17,         "2.0.x - 2.0.17");
```
Comment (asiekierka) [inline src/entry_names_gbdk.h:19]: I can't check what crt0 the pre-2.0.17 versions emitted due to missing builds, which is why this is used as a fallback.

File: src/entry_names_gbdk.h [hunk 2]
```diff
@@ -14,7 +14,12 @@
     // Tool
     DEF_NAME_STR(STR_GBDK, "GBDK");
     // Versions
-    DEF_NAME_STR(STR_GBDK_2_x_to_2020_3_2_0,       "2.x - 2020.3.2.0");
+    DEF_NAME_STR(STR_GBDK_1_x,                     "1.x");
```
Comment (asiekierka) [inline src/entry_names_gbdk.h]: Note that `gbtoolsid` did not detect GBDK 1.x before. This distinction may be useful for a changelog, though I'm not sure if anything used it "in production" other than the bundled examples.

File: src/sig_gbdk.c [hunk 3]
```diff
@@ -31,7 +31,34 @@ bool check_gbdk(void) {
...
+
+        // GBDK 2.0.18 - GBDK 2.0.15
+        if (CHECK_PATTERN_AT_ADDR(sig_gbdk_0x1c2_GBDK_2018_to_215, sig_gbdk_0x1c2_GBDK_2018_to_215_at)) {
+            entry_add_with_version(entry, STR_GBDK_2_0_18_to_2_1_5);
+            return true;
+        }
+
+        // GBDK 2.x, cannot narrow down further
+        // (for example, Demotronic uses a modified crt0 which fails the above checks)
+        entry_add_with_version(entry, STR_GBDK_2_x);
```
Comment (asiekierka) [inline src/sig_gbdk.c:67]: Not too sure what to do about this here.

File: src/entry_names_gbdk.h (hunk 1 above)
Comment (bbbbbr) [reply src/entry_names_gbdk.h:19]: Looks good, that's still much improved coverage. :+1:

File: src/sig_gbdk.c (hunk 3 above)
Comment (bbbbbr) [reply src/sig_gbdk.c:67]: That's ok, the same happens with modern GBDK-2020 programs that modify crt0. The modern ones, at least, aren't common and it just means they'll be detected as best they can. This tool is meant to work as a reasonably best-effort in general, as opposed to trying to capture the harder edge cases.

**Comment by bbbbbr** [comment]
I love that you added this!

**Pull request merged**